frontend/catalyst/from_plxpr.py

```diff
@@ -0,0 +1,380 @@
+    
+    If a value is not present in this dictionary, it needs to be extracted
+    from the ``qreg`` property.
+    """
+
+    op_math_cache: dict
+    """A cache of operations that will be consumed by later operations.
+    This is a map from the ``AbstractOperator`` variables to the corresponding
+    equation. The equation will need to be interpreted when the abstract
+    operator is consumed.
+    """
+
+    def __init__(self, device):
+        self._device = device
+        self.qreg = None
+        self.env = {}
+        self.wire_map = {}
+        self.op_math_cache = {}
+
+    def setup(self):
+        """Perform any customized setup and processing before processing the plxpr.
+
+        For conversion to catalyst, this allocates the device, extracts a register, and
+        resets the wire map.
+        """
+        qdevice_p.bind(**_get_device_kwargs(self._device))
+        self.qreg = qalloc_p.bind(len(self._device.wires))
+        self.wire_map = {}
+
+    def cleanup(self):
+        """Perform any final steps after processing the plxpr.
+
+        For conversion to calayst, this reinserts extracted qubits and
+        deallocates the register.
+        """
+        for orig_wire, wire in self.wire_map.items():
+            self.qreg = qinsert_p.bind(self.qreg, orig_wire, wire)
+        qdealloc_p.bind(self.qreg)
+
+    def read(self, var):
+        """Extract the value corresponding to a variable."""
+        return var.val if type(var) is jax.core.Literal else self.env[var]
+
+    def _get_wire(self, wire_value) -> AbstractQbit:
+        """Get the ``AbstractQbit`` corresponding to a wire value."""
+        if wire_value in self.wire_map:
+            return self.wire_map[wire_value]
+        return qextract_p.bind(self.qreg, wire_value)
+
+    def interpret_operator_eqn(self, eqn: jax.core.JaxprEqn) -> None:
+        """Interpret a plxpr equation describing an operation as a catalxpr equation."""
+        if not isinstance(eqn.outvars[0], jax.core.DropVar):
+            self.op_math_cache[eqn.outvars[0]] = eqn
+            return
+
+        if "n_wires" not in eqn.params:
+            raise NotImplementedError(
+                f"Operator {eqn.primitive.name} not yet supported for catalyst conversion."
+            )
+        n_wires = eqn.params["n_wires"]
+        split = None if n_wires == 0 else -n_wires
+
+        wire_values = [self.read(w) for w in eqn.invars[split:]]
+        wires = [self._get_wire(w) for w in wire_values]
+
+        invals = [self.read(invar) for invar in eqn.invars[:split]]
+
+        kwargs = {
+            "qubits_len": eqn.params["n_wires"],
+            "ctrl_len": 0,
+            "adjoint": False,
+        }
+
+        if eqn.primitive.name == "QubitUnitary":
+            outvals = qunitary_p.bind(*invals, *wires, **kwargs)
+        elif eqn.primitive.name == "GlobalPhase":
+            outvals = gphase_p.bind(*invals, ctrl_len=0, adjoint=False)
```

Comment:
This appears to be untested, should be an easy fix to add the gate to the tests.

Reply:
added.